compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -511,66 +494,69 @@ public void visit(BLangFunction funcNode) {
     }
 
     @Override
-    public void visit(BLangBlockFunctionBody body) {
-        SymbolEnv funcBodyEnv = SymbolEnv.createFuncBodyEnv(body, env);
+    public void visit(BLangBlockFunctionBody body, AnalyzerData data) {
+        SymbolEnv funcBodyEnv = SymbolEnv.createFuncBodyEnv(body, data.env);
         int stmtCount = -1;
         for (BLangStatement stmt : body.stmts) {
             stmtCount++;
             boolean analyzedStmt = analyzeBlockStmtFollowingIfWithoutElse(stmt,
-                    stmtCount > 0 ? body.stmts.get(stmtCount - 1) : null, funcBodyEnv, true);
+                    stmtCount > 0 ? body.stmts.get(stmtCount - 1) : null, funcBodyEnv, true, data);
             if (analyzedStmt) {
                 continue;
             }
-            analyzeStmt(stmt, funcBodyEnv);
+            data.env = funcBodyEnv;
+            analyzeStmt(stmt, data);
         }
         // Remove explicitly added function body env if exists
-        this.prevEnvs.remove(funcBodyEnv);
-        resetNotCompletedNormally();
+        data.prevEnvs.remove(funcBodyEnv);
+        resetNotCompletedNormally(data);
     }
 
     private boolean analyzeBlockStmtFollowingIfWithoutElse(BLangStatement currentStmt, BLangStatement prevStatement,
-                                                           SymbolEnv currentEnv, boolean inFunctionBody) {
+                                                           SymbolEnv currentEnv, boolean inFunctionBody,
+                                                           AnalyzerData data) {
```

Comment:
At the places where we call this method, `currentEnv` is always `data.env`, So do we need this argument?

Reply:
In L503 we need to pass new env. right?

Reply:
Ah yes, I missed that.